README.md
# Just a tag.

This crate contains the `Tag` type, an [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035)
DNS label compatible string, with parsing `FromStr` and optional [serde](https://serde.rs/) support.


## Tag examples

```rust
use justatag::Tag;

fn tags() {
    assert_eq!(Tag::new("some-tag"), "some-tag");
    assert_eq!(Tag::from_str("some-tag").unwrap(), "some-tag");
    assert!(Tag::from_str("invalid-").is_err());
}
```

## Unions of tags

A bit untrue to the crate's name, it also provides the `TagUnion` type, which represents
(unsurprisingly, this time) a union of tags.

```rust
use std::collections::HashSet;
use justatag::{MatchesAnyTagUnion, Tag, TagUnion};

fn tag_unions() {
    let union = TagUnion::from_str("foo").unwrap();
    assert!(union.contains(&Tag::new("foo")));
    assert_eq!(union.len(), 1);

    let union = TagUnion::from_str("foo+bar").unwrap();
    assert!(union.contains(&Tag::new("foo")));
    assert!(union.contains(&Tag::new("bar")));
    assert_eq!(union.len(), 2);

    // TagUnions are particularly interesting when bundled up.
    let unions = vec![
        TagUnion::from_str("bar+baz").unwrap(),
        TagUnion::from_str("foo").unwrap()
    ];

    // foo matches
    let set_1 = HashSet::from_iter([Tag::new("foo"), Tag::new("bar")]);
    assert!(unions.matches_set(&set_1));

    // bar+baz matches
    let set_2 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar"), Tag::new("baz")]);
    assert!(unions.matches_set(&set_2));

    // none match
    let set_3 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar")]);
    assert!(!unions.matches_set(&set_3));
}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
This project uses [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.1] - 2023-07-18

### Internal

- Fixes the examples to build on docs.rs.

## 0.1.0 - 2023-07-18

### Added

- 🎉 Initial release adding the `Tag` and `TagUnion` types.

[0.1.1]: https://github.com/sunsided/just-a-tag/releases/tag/0.1.1


src/tag_union.rs
//! Provides the [`TagUnion`] type, a union of tags.

// SPDX-FileCopyrightText: Copyright 2023 Markus Mayer
// SPDX-License-Identifier: EUPL-1.2
// SPDX-FileType: SOURCE

use crate::{Tag, TagFromStringError};
#[cfg(feature = "serde")]
use serde::{de, Deserialize, Deserializer};
use std::borrow::Borrow;
use std::collections::HashSet;
use std::error::Error;
use std::fmt::{Display, Formatter};
use std::hash::{Hash, Hasher};
use std::iter::FromIterator;
use std::ops::Deref;
use std::str::FromStr;

/// A tag union, e.g. `foo` or `foo+bar+baz` (i.e. `foo` _and_ `bar` _and_ `baz`).
///
/// ```
/// use just_a_tag::{Tag, TagUnion};
///
/// let union = TagUnion::from_str("foo").unwrap();
/// assert!(union.contains(&Tag::new("foo")));
/// assert_eq!(union.len(), 1);
///
/// let union = TagUnion::from_str("foo+bar").unwrap();
/// assert!(union.contains(&Tag::new("foo")));
/// assert!(union.contains(&Tag::new("bar")));
/// assert_eq!(union.len(), 2);
///
/// assert!(TagUnion::from_str("foo bar").is_err());
/// ```
#[derive(Debug, Default, Clone, Eq, PartialEq)]
pub struct TagUnion(HashSet<Tag>);

impl TagUnion {
    /// Returns `true` if this tag union matches the value presented in the set.
    ///
    /// ```
    /// use std::collections::HashSet;
    /// use just_a_tag::{MatchesAnyTagUnion, Tag, TagUnion};
    ///
    /// let unions = vec![
    ///     TagUnion::from_str("foo").unwrap(),
    ///     TagUnion::from_str("bar+baz").unwrap()
    /// ];
    ///
    /// // foo, and bar+baz matches
    /// let set_1 = HashSet::from_iter([Tag::new("foo"), Tag::new("bar"), Tag::new("baz")]);
    /// assert!(unions.matches_set(&set_1));
    ///
    /// // bar+baz matches
    /// let set_2 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar"), Tag::new("baz")]);
    /// assert!(unions.matches_set(&set_2));
    ///
    /// // foo matches
    /// let set_3 = HashSet::from_iter([Tag::new("foo"), Tag::new("bar")]);
    /// assert!(unions.matches_set(&set_3));
    ///
    /// // none match
    /// let set_4 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar")]);
    /// assert!(!unions.matches_set(&set_4));
    /// ```
    pub fn matches_set(&self, values: &HashSet<Tag>) -> bool {
        self.0.is_subset(values)
    }

    /// Inserts a tag into this union.
    /// Returns whether the tag was inserted; that is:
    ///
    /// * If the tag was not previously inserted, `true` is returned,
    /// * If the tag was previously inserted, `false` is returned.
    pub fn insert(&mut self, tag: Tag) -> bool {
        self.0.insert(tag)
    }

    /// Removes a tag from this union.
    /// Returns whether the tag was removed; that is:
    ///
    /// * If the tag was previously inserted, `true` is returned,
    /// * If the tag was not previously inserted, `false` is returned.
    pub fn remove<T: Borrow<Tag>>(&mut self, tag: T) -> bool {
        self.0.remove(tag.borrow())
    }

    /// Returns whether this union contains the specified tag. That is:
    ///
    /// * If the tag was previously inserted, `true` is returned,
    /// * If the tag was not previously inserted, `false` is returned.
    pub fn contains<T: Borrow<Tag>>(&self, tag: &T) -> bool {
        self.0.contains(tag.borrow())
    }

    /// Attempts to parse a [`TagUnion`] from a string-like input.
    pub fn from_str<S: AsRef<str>>(value: S) -> Result<TagUnion, TagUnionFromStringError> {
        let value = value.as_ref();
        if value.is_empty() {
            return Ok(TagUnion::default());
        }

        let parts = value.split('+');
        let names: HashSet<String> = parts
            .filter(|&c| !c.contains('+'))
            .filter(|&c| !c.is_empty())
            .map(|c| c.into())
            .collect();

        if names.is_empty() {
            return Ok(TagUnion::default());
        }

        let mut tags = HashSet::new();
        for name in names.into_iter() {
            tags.insert(Tag::from_str(&name)?);
        }

        Ok(Self(tags))
    }
}

/// Implements
pub trait MatchesAnyTagUnion {
    /// Returns `true` if this tag union matches the value presented in the set.
    ///
    /// ```
    /// use std::collections::HashSet;
    /// use just_a_tag::{MatchesAnyTagUnion, Tag, TagUnion};
    ///
    /// let unions = vec![
    ///     TagUnion::from_str("foo").unwrap(),
    ///     TagUnion::from_str("bar+baz").unwrap()
    /// ];
    ///
    /// // foo, and bar+baz matches
    /// let set_1 = HashSet::from_iter([Tag::new("foo"), Tag::new("bar"), Tag::new("baz")]);
    /// assert!(unions.matches_set(&set_1));
    ///
    /// // bar+baz matches
    /// let set_2 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar"), Tag::new("baz")]);
    /// assert!(unions.matches_set(&set_2));
    ///
    /// // foo matches
    /// let set_3 = HashSet::from_iter([Tag::new("foo"), Tag::new("bar")]);
    /// assert!(unions.matches_set(&set_3));
    ///
    /// // none match
    /// let set_4 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar")]);
    /// assert!(!unions.matches_set(&set_4));
    /// ```
    fn matches_set(&self, values: &HashSet<Tag>) -> bool;
}

impl MatchesAnyTagUnion for Vec<TagUnion> {
    fn matches_set(&self, values: &HashSet<Tag>) -> bool {
        self.iter().any(|s| s.matches_set(&values))
    }
}

impl Deref for TagUnion {
    type Target = HashSet<Tag>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Hash for TagUnion {
    fn hash<H: Hasher>(&self, state: &mut H) {
        let mut vec = Vec::from_iter(self.0.iter());
        vec.sort();
        for tag in vec {
            tag.hash(state);
        }
    }
}

impl FromIterator<Tag> for TagUnion {
    fn from_iter<T: IntoIterator<Item = Tag>>(iter: T) -> Self {
        Self(iter.into_iter().collect())
    }
}

impl FromStr for TagUnion {
    type Err = TagUnionFromStringError;

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        TagUnion::from_str(value)
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for TagUnion {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let input = String::deserialize(deserializer)?;
        match TagUnion::from_str(&input) {
            Ok(tags) => Ok(tags),
            Err(e) => Err(de::Error::custom(e)),
        }
    }
}

#[derive(Debug, Eq, PartialEq)]
pub enum TagUnionFromStringError {
    InvalidTag(TagFromStringError),
}

impl Display for TagUnionFromStringError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TagUnionFromStringError::InvalidTag(e) => write!(f, "Invalid tag: {e}"),
        }
    }
}

impl From<TagFromStringError> for TagUnionFromStringError {
    fn from(value: TagFromStringError) -> Self {
        Self::InvalidTag(value)
    }
}

impl Error for TagUnionFromStringError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty() {
        let tags: TagUnion = TagUnion::from_str("").unwrap();
        assert!(tags.is_empty());
    }

    #[test]
    fn test_add_remove() {
        let mut tags = TagUnion::from_str(r#"foo"#).unwrap();
        assert!(tags.contains(&Tag::new("foo")));
        assert_eq!(tags.len(), 1);

        tags.insert(Tag::new("bar"));
        assert_eq!(tags.len(), 2);
        tags.contains(&Tag::new("bar"));

        tags.remove(&Tag::new("foo"));
        assert!(!tags.contains(&Tag::new("foo")));
        assert_eq!(tags.len(), 1);

        tags.remove(&Tag::new("bar"));
        assert_eq!(tags.len(), 0);
        assert!(tags.is_empty());
    }

    #[test]
    #[cfg(feature = "serde")]
    fn test_trivial() {
        let tags: TagUnion = serde_json::from_str(r#""foo""#).unwrap();
        assert!(tags.contains(&Tag::new("foo")));
    }

    #[test]
    #[cfg(feature = "serde")]
    fn test_complex() {
        let tags: TagUnion = serde_json::from_str(r#""foo+bar+++baz++""#).unwrap();
        assert_eq!(tags.len(), 3);
        assert!(tags.contains(&Tag::new("foo")));
        assert!(tags.contains(&Tag::new("bar")));
        assert!(tags.contains(&Tag::new("baz")));
    }

    #[test]
    fn test_invalid() {
        let tags = TagUnion::from_str(r#"foo+#baz"#);
        assert_eq!(
            tags,
            Err(TagUnionFromStringError::InvalidTag(
                crate::TagFromStringError::MustStartAlphabetic('#')
            ))
        );
    }

    #[test]
    fn test_matches() {
        let selections = vec![
            TagUnion::from_str("foo+bar").unwrap(),
            TagUnion::from_str("baz").unwrap(),
        ];

        // foo+bar are present, so is baz
        assert!(selections.matches_set(&HashSet::from_iter([
            Tag::new("foo"),
            Tag::new("bar"),
            Tag::new("baz"),
        ])));

        // baz is present
        assert!(selections.matches_set(&HashSet::from_iter([Tag::new("baz"),])));

        // foo+bar are present
        assert!(selections.matches_set(&HashSet::from_iter([Tag::new("foo"), Tag::new("bar"),])));

        // baz present
        assert!(selections.matches_set(&HashSet::from_iter([Tag::new("foo"), Tag::new("baz"),])));

        // neither foo+bar, nor baz are present.
        assert!(!selections.matches_set(&HashSet::from_iter([Tag::new("foo"), Tag::new("bang"),])));
    }
}


src/lib.rs
//! # Just a tag
//!
//! This crate contains the [`Tag`] type, an [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035)
//! DNS label compatible string, with parsing [`FromStr`] and optional [serde](https://serde.rs/) support.
//!
//! ## Tag examples
//!
//! ```
//! # use just_a_tag::Tag;
//! assert_eq!(Tag::new("some-tag"), "some-tag");
//! assert_eq!(Tag::from_str("some-tag").unwrap(), "some-tag");
//! assert!(Tag::from_str("invalid-").is_err());
//! ```
//!
//! ## Unions of tags
//!
//! A bit untrue to the crate's name, it also provides the [`TagUnion`] type, which represents
//! (unsurprisingly, this time) a union of tags.
//!
//! ```
//! use std::collections::HashSet;
//! use just_a_tag::{MatchesAnyTagUnion, Tag, TagUnion};
//!
//! let union = TagUnion::from_str("foo").unwrap();
//! assert!(union.contains(&Tag::new("foo")));
//! assert_eq!(union.len(), 1);
//!
//! let union = TagUnion::from_str("foo+bar").unwrap();
//! assert!(union.contains(&Tag::new("foo")));
//! assert!(union.contains(&Tag::new("bar")));
//! assert_eq!(union.len(), 2);
//!
//! // TagUnions are particularly interesting when bundled up.
//! let unions = vec![
//!     TagUnion::from_str("bar+baz").unwrap(),
//!     TagUnion::from_str("foo").unwrap()
//! ];
//!
//! // foo matches
//! let set_1 = HashSet::from_iter([Tag::new("foo"), Tag::new("bar")]);
//! assert!(unions.matches_set(&set_1));
//!
//! // bar+baz matches
//! let set_2 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar"), Tag::new("baz")]);
//! assert!(unions.matches_set(&set_2));
//!
//! // none match
//! let set_3 = HashSet::from_iter([Tag::new("fubar"), Tag::new("bar")]);
//! assert!(!unions.matches_set(&set_3));
//! ```

// SPDX-FileCopyrightText: Copyright 2023 Markus Mayer
// SPDX-License-Identifier: EUPL-1.2
// SPDX-FileType: SOURCE

// Only enable the `doc_cfg` feature when the `docsrs` configuration attribute is defined.
#![cfg_attr(docsrs, feature(doc_cfg))]
#![cfg_attr(feature = "unsafe", allow(unsafe_code))]
#![cfg_attr(not(feature = "unsafe"), forbid(unsafe_code))]

mod tag_union;

#[cfg(feature = "serde")]
use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
use std::error::Error;
use std::fmt::{Display, Formatter};
use std::ops::Deref;
use std::str::FromStr;

pub use tag_union::{MatchesAnyTagUnion, TagUnion, TagUnionFromStringError};

/// A tag name.
///
/// Tag names [RFC 1035](https://datatracker.ietf.org/doc/html/rfc1035) DNS label compatible,
/// in other words they must
///
/// - not be longer than 63 characters,,
/// - only lowercase alphanumeric characters or '-',
/// - start with an alphabetic character, and
/// - end with an alphanumeric character.
#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Tag(String);

impl Tag {
    /// An empty tag.
    pub const EMPTY: Tag = Tag(String::new());

    /// The maximum length of a tag.
    pub const MAX_LEN: usize = 63;

    /// Constructs a new tag.
    ///
    /// ## Panics
    ///
    /// This method panics if the input is not a valid tag. If you want to avoid a panic,
    /// use [`Tag::from_str`](Self::from_str) instead.
    ///
    /// ## Example
    ///
    /// ```
    /// use just_a_tag::Tag;
    /// assert_eq!(Tag::new("foo"), "foo");
    /// ```
    pub fn new<V: AsRef<str>>(value: V) -> Self {
        value.as_ref().parse().expect("invalid input")
    }

    /// Constructs a new tag without checking for validity.
    ///
    /// ## Example
    ///
    /// ```
    /// # use just_a_tag::Tag;
    /// /// // Constructs a Tag without verifying the input.
    /// assert_eq!(unsafe { Tag::new_unchecked("foo") }, "foo");
    /// assert_eq!(unsafe { Tag::new_unchecked("@") }, "@"); // NOTE: invalid input
    /// ```
    #[cfg_attr(docsrs, doc(cfg(feature = "unsafe")))]
    #[cfg(feature = "unsafe")]
    pub unsafe fn new_unchecked<V: Into<String>>(value: V) -> Self {
        Self(value.into())
    }

    /// Parses a [`Tag`] from a string-like value.
    ///
    /// ```
    /// # use just_a_tag::Tag;
    /// assert_eq!(Tag::from_str("some-tag").unwrap(), "some-tag");
    /// assert!(Tag::from_str("invalid-").is_err());
    /// ```
    pub fn from_str<S: AsRef<str>>(value: S) -> Result<Self, TagFromStringError> {
        let value = value.as_ref();
        if value.is_empty() {
            return Ok(Tag::EMPTY.clone());
        }

        if value.len() > Tag::MAX_LEN {
            return Err(TagFromStringError::LimitExceeded(value.len()));
        }

        let mut chars = value.chars();
        let first = chars.next().expect("tag is not empty");
        if !first.is_ascii_lowercase() {
            return Err(TagFromStringError::MustStartAlphabetic(first));
        }

        let mut previous = first;
        while let Some(c) = chars.next() {
            if !c.is_ascii_digit() && !c.is_ascii_lowercase() && c != '-' {
                return Err(TagFromStringError::InvalidCharacter(c));
            }

            previous = c;
        }

        if !previous.is_ascii_lowercase() {
            return Err(TagFromStringError::MustEndAlphanumeric(previous));
        }

        Ok(Self(value.into()))
    }
}

impl Display for Tag {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Deref for Tag {
    type Target = str;

    #[inline(always)]
    fn deref(&self) -> &Self::Target {
        self.0.deref()
    }
}

impl PartialEq<str> for Tag {
    #[inline(always)]
    fn eq(&self, other: &str) -> bool {
        self.0.eq(other)
    }
}

impl PartialEq<&str> for Tag {
    #[inline(always)]
    fn eq(&self, other: &&str) -> bool {
        self.0.eq(other)
    }
}

impl FromStr for Tag {
    type Err = TagFromStringError;

    #[inline(always)]
    fn from_str(value: &str) -> Result<Self, Self::Err> {
        Tag::from_str(value)
    }
}

impl TryFrom<&str> for Tag {
    type Error = TagFromStringError;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        value.parse()
    }
}

impl TryFrom<String> for Tag {
    type Error = TagFromStringError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        value.parse()
    }
}

impl TryFrom<&String> for Tag {
    type Error = TagFromStringError;

    fn try_from(value: &String) -> Result<Self, Self::Error> {
        value.parse()
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for Tag {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let tag = String::deserialize(deserializer)?;
        match Tag::from_str(&tag) {
            Ok(tag) => Ok(tag),
            Err(e) => Err(de::Error::custom(e)),
        }
    }
}

#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
#[cfg(feature = "serde")]
impl Serialize for Tag {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(&self.0)
    }
}

#[derive(Debug, Eq, PartialEq)]
pub enum TagFromStringError {
    MustStartAlphabetic(char),
    MustEndAlphanumeric(char),
    InvalidCharacter(char),
    LimitExceeded(usize),
}

impl Display for TagFromStringError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            TagFromStringError::MustStartAlphabetic(c) => write!(
                f,
                "Tag name must begin with a lowercase alphabetic character, got '{c}'"
            ),
            TagFromStringError::MustEndAlphanumeric(c) => write!(
                f,
                "Tag name must end with a lowercase alphanumeric character, got '{c}'"
            ),
            TagFromStringError::InvalidCharacter(c) => write!(
                f,
                "Tag name must only contain lowercase alphanumeric characters or '-', got '{c}'"
            ),
            TagFromStringError::LimitExceeded(len) => write!(
                f,
                "Tag name must be not longer than 63 characters, got '{len}'"
            ),
        }
    }
}

impl Error for TagFromStringError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trivial() {
        assert_eq!(Tag::from_str("test").unwrap(), "test");
        assert_eq!(Tag::from_str("test-case").unwrap(), "test-case");
        assert_eq!(Tag::from_str("test---12e").unwrap(), "test---12e");
        assert!(
            Tag::from_str("a123456789a123456789a123456789a123456789a123456789a12345678901a")
                .is_ok()
        );
    }

    #[test]
    fn test_invalid() {
        assert!(Tag::from_str("1").is_err());
        assert!(Tag::from_str("-").is_err());
        assert!(Tag::from_str("a-").is_err());
        assert!(Tag::from_str("a1").is_err());
        assert!(Tag::from_str("a_b_c").is_err());
        assert!(
            Tag::from_str("a123456789a123456789a123456789a123456789a123456789a123456789012a")
                .is_err()
        );
    }

    #[test]
    #[cfg(feature = "serde")]
    fn test_serde_de_trivial() {
        assert_eq!(serde_json::from_str::<Tag>(r#""test""#).unwrap(), "test");
        assert_eq!(
            serde_json::from_str::<Tag>(r#""test-case""#).unwrap(),
            "test-case"
        );
        assert_eq!(
            serde_json::from_str::<Tag>(r#""test---12e""#).unwrap(),
            "test---12e"
        );
        assert!(serde_json::from_str::<Tag>(
            r#""a123456789a123456789a123456789a123456789a123456789a12345678901a""#
        )
        .is_ok());
    }

    #[test]
    #[cfg(feature = "serde")]
    fn test_serde_de_invalid() {
        assert!(serde_json::from_str::<Tag>(r#""1""#).is_err());
        assert!(serde_json::from_str::<Tag>(r#""-""#).is_err());
        assert!(serde_json::from_str::<Tag>(r#""a-""#).is_err());
        assert!(serde_json::from_str::<Tag>(r#""a1""#).is_err());
        assert!(serde_json::from_str::<Tag>(r#""a_b_c""#).is_err());
        assert!(serde_json::from_str::<Tag>(
            r#""a123456789a123456789a123456789a123456789a123456789a123456789012a""#
        )
        .is_err());
    }

    #[test]
    #[cfg(feature = "serde")]
    fn test_serde_ser_invalid() {
        let json = serde_json::to_string(&Tag::new("foo")).unwrap();
        assert_eq!(json, r#""foo""#);
    }
}
